rfd/0078-role-variables.md

```diff
@@ -0,0 +1,176 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 78 - Role Variables
+
+## What
+
+Teleport admins should be able to define variables within roles which can extend
+claims coming from internal and external user traits.
+
+Requirements:
+
+1. Variable definitions should support some form of pattern matching so that
+   traits can be easily and powerfully extended.
+2. A variable defined in one role should be usable in many other roles.
+
+Non-requirements:
+
+1. Variables can depend on other variables.
+
+## Why
+
+Variables will be useful to specify additional traits that are not available in
+OIDC/SAML claims.
+
+Teleport admins don't always have control over their org's OIDC/SAML provider.
+
+Role variables will simplify role evaluation, keep logic in one place, and
+minimize repetition.
+
+## Details
+
+Example role spec:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: node-users
+spec:
+  vars:
+    - name: "logins"
+      # "select" is a string value which will be used as the input to all
+      # possible expansion matchers. It can (and usually will) contain a trait
+      # which will be expanded before passing as input.
+      select: "{{external.email}}"
```

Comment:
I guess the ship has sailed with the `{{` `}}` syntax since we already use it for traits. I wish we would have gone with something different, because if you forget to quote it you get invalid YAML.

Reply:
Yeah I was planning to just use the regular trait expansion here, but we could make it so you give the name only, like `select: external.email`. That might make it more clear that there can only be 1, and and prefix/suffix is pointless anyway, I will probably switch to that

Reply:
Never mind, I will probably leave the `{{}}` syntax as-is to keep things simple and consistent with other uses of traits.